dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/bundle/DefaultTrackerBundleService.java

```diff
@@ -442,90 +439,41 @@ private TrackerTypeReport handleRelationships( Session session, TrackerBundle bu
-                {
-                    unassignedFileResources.add( attributeValueMap.get( at.getAttribute() ).getValue() );
-                }
-            }
-
             TrackedEntityAttribute attribute = preheat.get( TrackerIdScheme.UID, TrackedEntityAttribute.class,
                 at.getAttribute() );
-            // TODO: What to do here? Should attribute == NULL this be allowed?
-            TrackedEntityAttributeValue attributeValue = null;
-
-            if ( attributeValueMap.containsKey( at.getAttribute() ) )
-            {
-                TrackedEntityAttributeValue av = attributeValueMap.get( at.getAttribute() );
-
-                av.setAttribute( attribute ).setValue( at.getValue() ).setStoredBy( at.getStoredBy() );
 
-                attributeValue = av;
-                attributeValues.add( attributeValue );
-            }
-
-            // new attribute value
-            if ( attributeValue == null )
-            {
-                attributeValue = new TrackedEntityAttributeValue();
-
-                attributeValue.setAttribute( attribute ).setValue( at.getValue() ).setStoredBy( at.getStoredBy() );
-
-                attributeValues.add( attributeValue );
-            }
-
-            checkNotNull( attributeValue.getAttribute(),
+            checkNotNull( attribute,
                 "Attribute should never be NULL here if validation is enforced before commit." );
 
-            // TODO: What to do here? Should this be allowed? i.e ,  attributeValue.getAttribute() != null  this makes a NP
-            if ( !attributeValuesForDeletion.contains( at.getAttribute() ) &&
-                attributeValue.getAttribute().getValueType().isFile() )
-            {
-                assignedFileResources.add( at.getValue() );
-            }
-        }
+            TrackedEntityAttributeValue attributeValue = attributeValueDBMap.getOrDefault( at.getAttribute(),
+                new TrackedEntityAttributeValue() );
 
-        for ( TrackedEntityAttributeValue attributeValue : attributeValues )
-        {
-            // since TEAV is the owning side here, we don't bother updating the TE.teav
-            // collection
-            // as it will be reloaded on session clear
-            TrackedEntityAttribute attribute = attributeValue.getAttribute();
+            attributeValue
+                .setAttribute( attribute )
+                .setEntityInstance( trackedEntityInstance )
+                .setValue( at.getValue() )
+                .setStoredBy( at.getStoredBy() );
 
-            checkNotNull( attribute,
-                "Attribute should never be NULL here if validation is enforced before commit." );
-
-            if ( attributeValuesForDeletion.contains( attribute.getUid() ) )
+            // We cannot use attributeValue.getValue() because it uses encryption logic
+            // So we need to use at.getValue()
+            if ( StringUtils.isEmpty( at.getValue() ) )
             {
```

Comment:
I wonder if it would not be more efficient to handle the file resources assignment by first checking if the Data Value is linked to a Data Element of type `FILE_RESOURCE`

Reply:
Absolutely, it got lost in the refactor